agent/src/main/java/com/microsoft/applicationinsights/agent/internal/coresync/impl/ImplementationsCoordinator.java

```diff
@@ -0,0 +1,171 @@
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.agent.internal.coresync.impl;
+
+import java.net.URL;
+import java.sql.Statement;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.microsoft.applicationinsights.agent.internal.agent.StringUtils;
+import com.microsoft.applicationinsights.agent.internal.logger.InternalAgentLogger;
+import com.microsoft.applicationinsights.agent.internal.coresync.AgentNotificationsHandler;
+
+/**
+ * The class serves as the contact point between injected code and its real implementation.
+ * Injected code will notify on various events in the class, calling this class, the class
+ * will then delegate the call to the relevant 'client' by consulting the data on the Thread's TLS
+ *
+ * Note that the called methods should not throw any exception or error otherwise that might affect user's code
+ *
+ * Created by gupele on 5/6/2015.
+ */
+public enum ImplementationsCoordinator implements AgentNotificationsHandler {
+    INSTANCE;
+
+    @Override
+    public void onException(String className, String methodName, Throwable throwable) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onException(className, methodName, throwable);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodEnterURL(String name, URL url) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                implementation.onMethodEnterURL(name, url);
+            }
+        } catch (Throwable t) {
+        }
+    }
+
+    @Override
+    public void onMethodEnterSqlStatement(String name, Statement statement, String sqlStatement) {
+        try {
+            AgentNotificationsHandler implementation = getImplementation();
+            if (implementation != null) {
+                if (statement != null && sqlStatement != null && sqlStatement.length() == 0) {
```

Comment:
Wouldn't you want to set `sqlStatement` also in case that `sqlStatement == null`?